Add passphrase input to the add SSH key dialog

vinokurig opened the pull request:
<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests

COMMITTERS: please include labels on each PR. Labels are listed here: https://github.com/eclipse/che/wiki/Labels but at a minimum you should include `kind/..` and `Dev Open Pull Request Status` labels.
-->

### What does this PR do?
**DO NOT MERGE**

* Add a new input for `passphrase` value to the add SSH key dialog:
![screenshot-eclipse-che apps rosa kd8yx-tndh6-rve jav8 p3 openshiftapps com-2024 08 01-09_52_48](https://github.com/user-attachments/assets/2d497a19-0518-4d9f-8762-088dbc1ccddf)
* Propagate the new `passphrase` value to the ssh key secret.

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/CRW-6614

### Is it tested? How?
<!-- 
Please provide instructions here which scenario you fix/implement
and in which way you tested it, provide as much as you think reviewer
needs to do the same.
-->
1. Go to User Preferences -> Ssh Keys tab
2. Fill the public and private key inputs, enter some value to the `passphrase` input.
3. Save the key and get the `git-ssh-key` secret data.

See: the `passphrase` field is present in the data map and its value equals to the input.

#### Release Notes
<!-- markdown to be included in marketing announcement - N/A for bugs -->


#### Docs PR
<!-- Please add a matching PR to [the docs repo](https://github.com/eclipse/che-docs) and link that PR to this issue.
Both will be merged at the same time. -->

akurinnoy commented:
@vinokurig can you please take a look at why the build failed?

dmytro-ndp commented:
@vinokurig : could you. please, confirm, that it should be enough to install Eclipse Che Next with User Dashboard deployed from quay.io/eclipse/che-dashboard:pr-1157 to validate the PR?

Asking, because the message in PR description mentioned https://github.com/devfile/devworkspace-operator/pull/1291, so I am curious if it's needed to have specific **devworkspace-operator** as well.

AObuchow commented:
> Asking, because the message in PR description mentioned [devfile/devworkspace-operator#1291](https://github.com/devfile/devworkspace-operator/pull/1291), so I am curious if it's needed to have specific **devworkspace-operator** as well.

@dmytro-ndp For a full test flow in Che, I believe you'll need to use the DWO project clone container built from https://github.com/devfile/devworkspace-operator/pull/1291

I've already built this image and pushed it here: `quay.io/aobuchow/project-clone:ihor-sshkey-pr`

To make DWO use it, you'll need to configure your DWOC as follows:

```diff
kind: DevWorkspaceOperatorConfig
config:
  workspace:
+    projectClone:
+      image: quay.io/aobuchow/project-clone:ihor-sshkey-pr
```
Then create a new workspace for it to use the newly configured project clone container.

This _should_ work with the che-owned DWOC IIRC.

AObuchow commented:
@dmytro-ndp sorry, my above comment was wrong. What I described was testing the entire flow of the ssh passphrase feature.

Based on @vinokurig's PR description:
> See: the `passphrase` field is present in the data map and its value equals to the input.

I believe there should be a secret created in the user's namespace regarding the ssh key, and it should contain a passphrase field in it's data. Verifying that does not require a modified version of DWO or the project clone container.

vinokurig commented:
@dmytro-ndp 
> @vinokurig : could you. please, confirm, that it should be enough to install Eclipse Che Next with User Dashboard deployed from quay.io/eclipse/che-dashboard:pr-1157 to validate the PR?

As @AObuchow mentioned in his comment above we do not need the modified DWO image here. In the scope of the pull request it would be enough to verify that the `Add SSH key` dialog has the `Passphrase` input and that the ssh-secret created after adding the SSH key has the `passphrase` data.

dmytro-ndp commented:
Thank you for the explanation, @AObuchow , @vinokurig !

dmytro-ndp commented:
@vinokurig: I faced error `Failed to add the SSH key. Unable to add SSH key "git-ssh-key": Secret in version "v1" cannot be handled as a Secret: illegal base64 data at input byte 4` when executed step "_3. Save the key and get the git-ssh-key secret data._" with SSH key of size 2048:
![Screenshot from 2024-08-05 21-54-32](https://github.com/user-attachments/assets/49687b64-2581-4e2a-b50f-3ad6cc59d5cf)
[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.08.05-21_47_01.webm](https://github.com/user-attachments/assets/fd7044ce-4c02-463c-9eb5-ea43d0668aad)

**_Test environment_**: OCP 4.16 cluster with Eclipse Che Next. using User Dashboard quay.io/eclipse/che-dashboard:pr-1157 , and DWOC quay.io/devfile/devworkspace-controller:sha-e176ec0

SSH key were created by the command: `ssh-keygen -b 2048 -t rsa -N "123qwe"`

It would be great if you could take a look.

vinokurig commented:
@dmytro-ndp Thank you for reporting the error. I fixed the error in the latest commit. Also I changed the passphrase input type to password so the sensitive data is hidden:
![screenshot-eclipse-che apps rosa gwoee-g8jth-qqt 1742 p3 openshiftapps com-2024 08 06-10_21_31](https://github.com/user-attachments/assets/3c9645de-af96-4cbc-b5f8-eefee2eb9a11)

dmytro-ndp commented:
@vinokurig : one more notice: Uses Dashboard failed to add SSH key with passphrase containing Cyrillic symbols: `Failed to add the SSH key. Unable to add SSH key "git-ssh-key": Invalid character`
![Screenshot from 2024-08-06 15-14-05](https://github.com/user-attachments/assets/0737b8c4-3027-4275-89a8-c9f203c81b62)

SSH key were created using the command `ssh-keygen -b 2048 -t rsa -N "тест"`.

vinokurig commented:
@dmytro-ndp Thank you for the deep testing and reporting the bugs. Despite that fact that the bugs are not related to the pull request, I think we should fix the bugs before merging the pull request. It is my bad that I did not try some git operations in the scope of the [DWO pull request](https://github.com/devfile/devworkspace-operator/pull/1291). 
@dmytro-ndp Could you please create an issue for the found bugs?

dmytro-ndp commented:
@vinokurig , @AObuchow : here is the issue about pushing changes to the git repo: https://github.com/devfile/devworkspace-operator/issues/1295

ibuziuk commented:
In this case I propose to postpone this feature and move it to 3.17 / 7.91

vinokurig commented:
@dmytro-ndp The SSH flow works for both git UI and terminal commands, could you please validate the pull request again?

dmytro-ndp reviewed (approved):
@vinokurig : fixed version Eclipse Che Next + quay.io/eclipse/che-dashboard@sha256:ac07ba98b10ee8b40eedfa4b8aae51f084a6396a627abdad774de6004b9a89cc worked as expected with ssh key using passphrase:
1) push changes from the terminal in workspace
2) push changes using GitHub Pull extension

Well done!

Screencast: 

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.09.13-21_39_37.webm](https://github.com/user-attachments/assets/dd075eb0-af5b-4b0c-99c4-abacfcc39fb0)

olexii4 reviewed (approved):
LGTM

devstudio-release commented:
Build [3.17](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [dashboard_3.x/543](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/dashboard_3.x/543/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/dashboard_3.x/543/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/dashboard_3.x/543/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/dashboard_3.x/543/git)

devstudio-release commented:
Build [3.17](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [sync-to-downstream_3.x/7703](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/7703/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/7703/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/7703/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/7703/git)

devstudio-release commented:
Build [3.17](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [get-sources-rhpkg-container-build_3.x/7695](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/7695/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/7695/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/7695/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/7695/git)
